BaseClasses.py

```diff
@@ -98,34 +98,42 @@ def __getitem__(self, player) -> bool:
             return self.rule(player)
 
     class RegionManager:
```

Comment:
We have an opportunity here to make it so when a world gets a location/region/entrance out of this, type-checkers and IDEs can know which subclass it is, and know about properties that are on that game's object.

`    class RegionManager(Generic[_T_Reg, _T_Ent, _T_Loc]):`

in the world:

`    regions: "MultiWorld.RegionManager[MessengerRegion, MessengerEntrance, MessengerLocation]"`

Reply:
Wouldn't you still have to overwrite the getter methods to get proper typing out of those?

Reply:
If you're talking about `World.get_location` et al., yes. That's why I've never liked those getters and I wouldn't use them.

as expressed in the first bullet point here:
https://github.com/ArchipelagoMW/Archipelago/pull/3021#issuecomment-2040068339
(This PR is solving the 2nd bullet point.)

I think it would be better to deprecate those methods and put them on this class instead.
Another option would be to make the whole `World` class generic, but I don't think that would be worth all the noise on all the `World` type annotations throughout the code base, just to get something out of these few methods that aren't even used in that many places.